lib/ast-walk.js

```diff
@@ -0,0 +1,52 @@
+var parse = require('acorn').parse;
+
+module.exports = function (src, opts) {
+    if (!opts) opts = {}
+    var ast = src;
+    if (typeof src === 'string') {
+        try {
+            ast = parse(src, {
+                ecmaVersion: 9,
+                allowReturnOutsideFunction: true
+            })
+        }
+        catch (err) { ast = parse('(' + src + ')') }
+    }
+    return function (cb) {
+        walk(ast, undefined, cb);
+    };
+};
+
+function walk (node, parent, cb) {
+    var keys = objectKeys(node);
+    for (var i = 0; i < keys.length; i++) {
+        var key = keys[i];
+        if (key === 'parent') continue;
+        
+        var child = node[key];
+        if (isArray(child)) {
+            for (var j = 0; j < child.length; j++) {
+                var c = child[j];
+                if (c && typeof c.type === 'string') {
+                    c.parent = node;
+                    walk(c, node, cb);
+                }
+            }
+        }
+        else if (child && typeof child.type === 'string') {
+            child.parent = node;
+            walk(child, node, cb);
+        }
+    }
+    cb(node);
+}
+
+var isArray = Array.isArray || function (xs) {
```

Comment:
What version of ES does `Array.isArray` support start in? If its node 4, I don't think we need the shim.